File: rust/src/prefix.rs

// use core::fmt::Debug;
//use std::ops::{Add, Sub};
use num::bigint::BigUint;

use num_traits::identities::Zero;
use num_traits::identities::One;
use core::ops::Shl;
use core::ops::Add;
use core::cmp::Ordering;
use core::cmp::Ord;
// use num_integer::Integer;
use num_traits::cast::ToPrimitive;
// use num_traits::cast::FromPrimitive;

use std::fmt;

pub struct Prefix {
    pub num: usize,
    pub ip_bits: ::ip_bits::IpBits,
    pub net_mask: BigUint,
    pub vt_from: fn(&Prefix, usize) -> Result<Prefix, String>,
    //pub vt_to_ip_str: &'static (Fn(&Vec<u16>) -> String)
}

impl Clone for Prefix {
    fn clone(&self) -> Prefix {
        Prefix {
            num: self.num,
            ip_bits: self.ip_bits.clone(),
            net_mask: self.net_mask.clone(),
            vt_from: self.vt_from
        }
    }
}

impl PartialEq for Prefix {
    fn eq(&self, other: &Self) -> bool {
        return self.ip_bits.version == other.ip_bits.version &&
          self.num == other.num;
    }
    fn ne(&self, other: &Self) -> bool {
        !self.eq(other)
    }
}

// funny
impl fmt::Debug for Prefix {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Prefix: {}", self.num)
    }
}


impl Eq for Prefix {}

impl Ord for Prefix {
    fn cmp(&self, oth: & Prefix) -> Ordering {
        if self.ip_bits.version < oth.ip_bits.version {
            Ordering::Less
        } else if self.ip_bits.version > oth.ip_bits.version {
            Ordering::Greater
        } else {
            if self.num < oth.num {
                Ordering::Less
            } else if self.num > oth.num {
                Ordering::Greater
            } else {
                Ordering::Equal
            }
        }
    }
}
impl PartialOrd for Prefix {
    fn partial_cmp(&self, other: &Prefix) -> Option<Ordering> {
        Some(self.cmp(other))
    }

}

impl Prefix {
    //#[allow(dead_code)]
    pub fn from(&self, num: usize) -> Result<Prefix, String>{
        return (self.vt_from)(self, num)
    }

    #[allow(dead_code)]
    pub fn to_ip_str(&self) -> String {
        return (self.ip_bits.vt_as_compressed_string)(&self.ip_bits, &self.netmask())
    }

    #[allow(dead_code)]
    pub fn size(&self) -> BigUint {
      return BigUint::one() << (self.ip_bits.bits-self.num.to_usize().unwrap())
    }

    pub fn new_netmask(prefix: usize, bits: usize) -> BigUint {
        let mut mask = BigUint::zero();
        let host_prefix = bits-prefix;
        for i in 0..prefix {
            mask = mask + (BigUint::one() << (host_prefix+i));
        }
        return mask
    }

    #[allow(dead_code)]
    //#[allow(unused_variables)]
    pub fn netmask(&self) -> BigUint {
        self.net_mask.clone()
    }

    #[allow(dead_code)]
    pub fn get_prefix(&self) -> usize {
        return self.num
    }

    ///  The hostmask is the contrary of the subnet mask,
    ///  as it shows the bits that can change within the
    ///  hosts
    ///
    ///    prefix = IPAddress::Prefix32.new 24
    ///
    ///    prefix.hostmask
    ///      ///  "0.0.0.255"
    ///
    #[allow(dead_code)]
    pub fn host_mask(&self) -> BigUint {
        let mut ret = BigUint::zero();
        for _ in 0..(self.ip_bits.bits-self.num) {
            ret = ret.shl(1).add(BigUint::one());
        }
        return ret;
    }

    ///
    ///  Returns the length of the host portion
    ///  of a netmask.
    ///
    ///    prefix = Prefix128.new 96
    ///
    ///    prefix.host_prefix
    ///      ///  128
    ///
    #[allow(dead_code)]
    pub fn host_prefix(&self) -> usize {
        return (self.ip_bits.bits) - self.num;
    }

    ///
    ///  Transforms the prefix into a string of bits
    ///  representing the netmask
    ///
    ///    prefix = IPAddress::Prefix128.new 64
    ///
    ///    prefix.bits
    ///      ///  "1111111111111111111111111111111111111111111111111111111111111111"
    ///          "0000000000000000000000000000000000000000000000000000000000000000"
    ///
    #[allow(dead_code)]
    pub fn bits(&self) -> String {
        return self.netmask().to_str_radix(2)
    }
    #[allow(dead_code)]
    pub fn to_s(&self) -> String {
        return format!("{}", self.get_prefix());
    }
    #[allow(dead_code)]
    pub fn to_i(&self) -> usize {
        return self.get_prefix();
    }

    #[allow(dead_code)]
    pub fn add_prefix(&self, other: &Prefix) -> Result<Prefix, String> {
        self.from(self.get_prefix() + other.get_prefix())
    }
    #[allow(dead_code)]
    pub fn add(&self, other: usize) -> Result<Prefix, String> {
        self.from(self.get_prefix() + other)
    }
    #[allow(dead_code)]
    pub fn sub_prefix(&self, other: &Prefix) -> Result<Prefix, String> {
        return self.sub(other.get_prefix());
    }
    #[allow(dead_code)]
    pub fn sub(&self, other: usize) -> Result<Prefix, String> {
        if other > self.get_prefix() {
            return self.from(other-self.get_prefix());
        }
        return self.from(self.get_prefix() - other);
    }

}

File: js/src/prefix.ts

import IpBits from './ip_bits';
import Crunchy from './crunchy';

interface From {
    (source: Prefix, num: number): Prefix;
}


class Prefix {
    num: number;
    ip_bits: IpBits;
    net_mask: Crunchy;
    vt_from: From;

    constructor(obj: {[id:string] : any}) {
       this.num = obj['num']; 
       this.ip_bits = obj['ip_bits']; 
       this.net_mask = obj['net_mask']; 
       this.vt_from = obj['vt_from']; 
    }

    public clone() : Prefix {
        return new Prefix({
            num: this.num,
            ip_bits: this.ip_bits,
            net_mask: this.net_mask,
            vt_from: this.vt_from
        });
    }

    public eq(other: Prefix) : boolean {
        return this.ip_bits.version == other.ip_bits.version &&
          this.num == other.num;
    }
    public ne(other: Prefix) : boolean {
        return !this.eq(other);
    }
    public cmp(oth: Prefix) : number {
        if (this.ip_bits.version < oth.ip_bits.version) {
            return -1;
        } else if (this.ip_bits.version > oth.ip_bits.version) {
            return 1;
        } else {
            if (this.num < oth.num) {
                return -1;
            } else if (this.num > oth.num) {
                return 1;
            } else {
                return 0;
            }
        }
    }
    //#[allow(dead_code)]
    public from(num: number) : Prefix {
        return (this.vt_from)(this, num);
    }

    public to_ip_str() : string {
        return this.ip_bits.vt_as_compressed_string(this.ip_bits, this.net_mask);
    }

    public size() : Crunchy {
      return Crunchy.one().shl(this.ip_bits.bits-this.num);
    }

    public static new_netmask(prefix: number, bits: number) : Crunchy {
        let mask = Crunchy.zero();
        let host_prefix = bits-prefix;
        for (let i=0; i < prefix; ++i) {
            // console.log(">>>", i, host_prefix, mask);
            mask = mask.add(Crunchy.one().shl(host_prefix+i));
        }
        return mask
    }

    public netmask() : Crunchy {
        return this.net_mask;
    }

    public get_prefix() : number {
        return this.num;
    }

    //  The hostmask is the contrary of the subnet mask,
    //  as it shows the bits that can change within the
    //  hosts
    //
    //    prefix = IPAddress::Prefix32.new 24
    //
    //    prefix.hostmask
    //      // => "0.0.0.255"
    //
    public host_mask() : Crunchy {
        let ret = Crunchy.zero();
        for(let _ = 0; _ < (this.ip_bits.bits-this.num); _++) {
            ret = ret.shl(1).add(Crunchy.one());
        }
        return ret;
    }


    //
    //  Returns the length of the host portion
    //  of a netmask.
    //
    //    prefix = Prefix128.new 96
    //
    //    prefix.host_prefix
    //      // => 128
    //
    public host_prefix() : number {
        return this.ip_bits.bits - this.num;
    }

    //
    //  Transforms the prefix into a string of bits
    //  representing the netmask
    //
    //    prefix = IPAddress::Prefix128.new 64
    //
    //    prefix.bits
    //      // => "1111111111111111111111111111111111111111111111111111111111111111"
    //          "0000000000000000000000000000000000000000000000000000000000000000"
    //
    public bits() : string {
        return this.netmask().toString(2);
    }
    // #[allow(dead_code)]
    // public net_mask(&self) -> BigUint {
    //     return (self.in_mask.clone() >> (self.host_prefix() as usize)) << (self.host_prefix() as usize);
    // }

    public to_s() : string {
        return this.get_prefix().toString();
    }
    //#[allow(dead_code)]
    // public inspect(&self) -> String {
    //     return self.to_s();
    // }
    public to_i() : number {
        return this.get_prefix();
    }

    public add_prefix(other: Prefix) : Prefix {
        return this.from(this.get_prefix() + other.get_prefix());
    }

    public add(other: number) : Prefix {
        return this.from(this.get_prefix() + other)
    }

    public sub_prefix(other: Prefix) : Prefix {
        return this.sub(other.get_prefix());
    }

    public sub(other: number) : Prefix {
        if (other > this.get_prefix()) {
            return this.from(other-this.get_prefix());
        }
        return this.from(this.get_prefix() - other);
    }

}

export default Prefix;